### ISSUE TITLE
Bitbucket Server repo URLs are not case-sensitive

### ISSUE THREAD

**Issue by doxxx** [sourcegraph/sourcegraph#394]
- **Sourcegraph version:** Docker image using insiders tag (01208_2018-10-15_e694f9e)
- **Platform information:**
  - Ubuntu 16.04
  - Docker 18.06.1-ce
  - Bitbucket Server 5.13.1

#### Steps to reproduce:

1. Configure Sourcegraph to use Bitbucket Server instance.
2. Browse to a Go repo from that server in the Sourcegraph UI.
3. Hover over symbols which are contained in packages imported from a repo on the Bitbucket Server.

#### Expected behavior:

All symbols will have hover tooltips with valid information.

#### Actual behavior:

If the import path uses a lowercase Bitbucket Server project name, the tooltip contains an error: `Request languageFeatures/$provideHover failed with message: type/object not found`

The import path for a repo on a Bitbucket Server instance is `host/project/repo`, e.g. `git.synaptivemedical.com/alg/jobber`. This works in Go. However, the canonical URL actually uses uppercase for the project key, e.g. `git.synaptivemedical.com/ALG/jobber`. Bitbucket Server will accept either, but it appears that the Go language server uses a case-sensitive comparison.

Changing all the imports in the repo to use an uppercase project name resolves the issue, but I'm wondering if it's possible to make this change unnecessary?

**Comment by sqs** [comment]
Sorry, I categorized this issue wrong. It is (as you say) a problem in the Go lang server. Reassigned and relabeled.

**Comment by chrismwendt** [comment]
I attempted a repro using these two repos:

- https://github.com/chrismwendt/Foo with an [import comment](https://golang.org/cmd/go/#hdr-Import_path_checking) of `package Foo // import "github.com/chrismwendt/Foo"`
- https://github.com/chrismwendt/Bar which imports Foo using `import "github.com/chrismwendt/foo"` (notice it's not capitalized)

Based on my experimentation, Go rejects `go run main.go` in the Bar repo with the following error:

```
main.go:3:8: code in directory /Users/chrismwendt/go/src/github.com/chrismwendt/foo expects import "github.com/chrismwendt/Foo"
```

@doxxx, if this does not accurately represent your setup, I'll try again after you submit PRs to those repos or provide a minimal repro of your own.

It would also help if you could provide the following information in order to clarify the state of the project when the problem occurs:

- Your `package` statement (is it `package jobber // import "git.synaptivmedical.com/ALG/jobber"`?)
- Your `import` statement from the file that hovers are failing in (is it `git.synaptivemedical.com/alg/jobber`?)
- Where the token you're hovering over is defined (the same package, different package but same repo, or a different repo?)
- The name of the repository on your BitBucket instance

**Comment by chrismwendt** [comment]
Same here 🎄 

I did a bit more experimentation:

- When I remove the import comment, it runs successfully
- `go get github.com/chrismwendt/Bar` creates 2 directories `Bar` and `foo` under my `$GOPATH` (notice that `foo` is lower case, which matches the import statement in `Bar/main.go`)
- Code hosts (at least GitHub and Bitbucket) are able to find repos regardless of case (e.g. `git clone https://bitbucket.org/atlassian/stash-example-plugiN` works)

I've capitalized as much as I can think of: `package Foo` and the repo name is `chrismwendt/Foo`. However, `import "github.com/chrismwendt/foo"` in the Bar repo still works. I've exhausted my ideas for now, but let's see if anything's different in January.

**Comment by nicksnyder** [comment]
@chrismwendt If this is release blocking for 3.0, please add the release blocking label.

Otherwise, if this isn't going to get done in the next week, please remove it from the 3.0 milestone.

**Comment by doxxx** [comment]
@chrismwendt I"m back from vacation, so I'll be re-examining this situation today.

**Comment by doxxx** [comment]
TLDR: Sourcegraph is working correctly and I needed to fix my import paths.

This is what was happening:

I had a git clone of my repo at `$GOPATH/src/git.synaptivemedical.com/alg/jobber` and all the source in that repo used `git.synaptivemedical.com/alg/jobber` as the package prefix. I did not have any import comments.

Go was happy with this since everything in the `GOPATH` was consistent. But the canonical package name according to the way Bitbucket Server constructs its repo URLs is actually `git.synaptivemedical.com/ALG/jobber`. Given this, Sourcegraph was failing to find the lowercase package names mentioned in the source code.

I subsequently fixed all the import paths in the repo to use the uppercase form (renaming the folders in my local GOPATH as well) and Sourcegraph is able to find packages and show correct type information.

As further proof, if I try to `go get git.synaptivemedical.com/alg/jobber` using a temp `GOPATH` then I get a bunch of `case-insensitive import collision` errors from Go. If I try `go get git.synaptivemedical.com/ALG/jobber` then Go is happy. I hadn't tried this before, mea culpa.

Sorry for the confusion!

**Comment by sqs** [comment]
Thanks!
